VisualPinball.Unity/VisualPinball.Unity.Editor/Managers/ManagerWindow.cs

```diff
@@ -0,0 +1,288 @@
+			Engine.Math.Color val = EditorGUILayout.ColorField(new GUIContent(label, tooltip), field.ToUnityColor()).ToEngineColor();
+			if (EditorGUI.EndChangeCheck()) {
+				FinalizeChange(label, ref field, val);
+			}
+		}
+
+		protected void FinalizeChange<TField>(string label, ref TField field, TField val)
+		{
+			string undoName = "Edit " + DataTypeName + ": " + label;
+			OnDataChanged(undoName, _selectedItem);
+			Undo.RecordObject(_table, undoName);
+			field = val;
+			SceneView.RepaintAll();
+		}
+
+		protected string GetMemberValue(MemberInfo mi, object instance)
+		{
+			switch (mi) {
+				case FieldInfo fi: return fi.GetValue(instance) as string;
+				case PropertyInfo pi: return pi.GetValue(instance) as string;
+			}
+			return null;
+		}
+
+		protected bool IsReferenced(List<MemberInfo> mis, object instance, string refName)
+		{
+			if (mis == null) { return false; }
+			string refNameLower = refName.ToLower();
+			foreach (var mi in mis) {
+				if (GetMemberValue(mi, instance)?.ToLower() == refNameLower) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		protected void RenameReflectedFields(string undoName, IEditableItemBehavior item, List<MemberInfo> mis, string oldName, string newName)
+		{
+			foreach (var mi in mis) {
+				string fieldVal = GetMemberValue(mi, item.ItemData);
+				if (fieldVal == oldName) {
+					Undo.RecordObject(item as Object, undoName);
+					switch (mi) {
+						case FieldInfo fi: fi.SetValue(item.ItemData, newName); break;
+						case PropertyInfo pi: pi.SetValue(item.ItemData, newName); break;
+					}
+				}
+			}
+		}
+
+		private void UndoPerformed()
+		{
+			Reload();
+		}
+
+		private void ItemSelected(List<T> selectedItems)
+		{
+			_selectedItem = null;
+			if (selectedItems.Count > 0) {
+				_selectedItem = selectedItems[0]; // not supporting multi select for now
+				_renaming = false;
+			}
+			Repaint();
+		}
+
+		private void FindTable()
+		{
+			_table = GameObject.FindObjectOfType<TableBehavior>();
+			_data.Clear();
+			if (_table != null) {
+				CollectData(_data);
+			}
+			_listView = new ManagerListView<T>(_treeViewState, _data, ItemSelected);
+		}
+
+		private bool IsNameInUse(string name, T ignore = null)
+		{
+			foreach (var item in _data) {
+				if (item != ignore && name.ToLower() == item.Name.ToLower()) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		private string GetUniqueName(string desiredName, T ignore = null)
+		{
+			string acceptedName = desiredName;
+			int appendNum = 1;
+			while (IsNameInUse(acceptedName, ignore)) {
+				acceptedName = desiredName + appendNum;
+				appendNum++;
+			}
+			return acceptedName;
+		}
+
+		// check is a concrete class implements the given method name
+		private bool IsImplemented(string methodName)
+		{
+			var mi = this.GetType().GetMethod(methodName, BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
+			return mi != null && mi.GetBaseDefinition().DeclaringType != mi.DeclaringType;
+		}
```

Comment:
That's executed on each editor frame, no?

Reply:
True, but editor frames at least only happen when something changes. Either way a bad precedent, added caching.